In [155]:
import pandas as pd
import numpy as np

## DATA IMPORT

In [156]:
db = pd.read_csv('Resources/Top10000genre.csv')
top = pd.read_csv('Resources/Top200.csv')

## DATA CLEANING

In [157]:
db.head()

genre        artist_name                        track_name  \
0  Movie     Henri Salvador       C'est beau de faire un Show   
1  Movie  Martin & les fées  Perdu d'avance (par Gad Elmaleh)   
2  Movie    Joseph Williams    Don't Let Me Be Lonely Tonight   
3  Movie     Henri Salvador    Dis-moi Monsieur Gordon Cooper   
4  Movie       Fabien Nataf                         Ouverture   

                 track_id  popularity  acousticness  danceability  \
0  0BRjO6ga9RKCKjfDqeFgWV           0         0.611         0.389   
1  0BjC1NfoEOOusryehmNudP           1         0.246         0.590   
2  0CoSDzoNIKCRs124s9uTVy           3         0.952         0.663   
3  0Gc6TVm52BwZD07Ki6tIvf           0         0.703         0.240   
4  0IuslXpMROHdEPvSl1fTQK           4         0.950         0.331   

   duration_ms  energy  instrumentalness key  liveness  loudness   mode  \
0        99373   0.910             0.000  C#    0.3460    -1.828  Major   
1       137373   0.737             0.000  F#    0.1510    -5.559  Minor   
2       170267   0.131             0.000   C    0.1030   -13.879  Minor   
3       152427   0.326             0.000  C#    0.0985   -12.178  Major   
4        82625   0.225             0.123   F    0.2020   -21.150  Major   

   speechiness    tempo time_signature  valence  
0       0.0525  166.969            4/4    0.814  
1       0.0868  174.003            4/4    0.816  
2       0.0362   99.488            5/4    0.368  
3       0.0395  171.758            4/4    0.227  
4       0.0456  140.576            4/4    0.390

In [158]:
db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 232725 entries, 0 to 232724
Data columns (total 18 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   genre             232725 non-null  object 
 1   artist_name       232725 non-null  object 
 2   track_name        232725 non-null  object 
 3   track_id          232725 non-null  object 
 4   popularity        232725 non-null  int64  
 5   acousticness      232725 non-null  float64
 6   danceability      232725 non-null  float64
 7   duration_ms       232725 non-null  int64  
 8   energy            232725 non-null  float64
 9   instrumentalness  232725 non-null  float64
 10  key               232725 non-null  object 
 11  liveness          232725 non-null  float64
 12  loudness          232725 non-null  float64
 13  mode              232725 non-null  object 
 14  speechiness       232725 non-null  float64
 15  tempo             232725 non-null  float64
 16  time_signature    23

In [159]:
db.columns = db.columns.str.replace(' ', '')
db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 232725 entries, 0 to 232724
Data columns (total 18 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   genre             232725 non-null  object 
 1   artist_name       232725 non-null  object 
 2   track_name        232725 non-null  object 
 3   track_id          232725 non-null  object 
 4   popularity        232725 non-null  int64  
 5   acousticness      232725 non-null  float64
 6   danceability      232725 non-null  float64
 7   duration_ms       232725 non-null  int64  
 8   energy            232725 non-null  float64
 9   instrumentalness  232725 non-null  float64
 10  key               232725 non-null  object 
 11  liveness          232725 non-null  float64
 12  loudness          232725 non-null  float64
 13  mode              232725 non-null  object 
 14  speechiness       232725 non-null  float64
 15  tempo             232725 non-null  float64
 16  time_signature    23

In [160]:
top.head()

Index  Highest Charting Position  Number of Times Charted  \
0      1                          1                        8   
1      2                          2                        3   
2      3                          1                       11   
3      4                          3                        5   
4      5                          5                        1   

  Week of Highest Charting                          Song Name     Streams  \
0   2021-07-23--2021-07-30                            Beggin'  48,633,449   
1   2021-07-23--2021-07-30          STAY (with Justin Bieber)  47,248,719   
2   2021-06-25--2021-07-02                           good 4 u  40,162,559   
3   2021-07-02--2021-07-09                         Bad Habits  37,799,456   
4   2021-07-23--2021-07-30  INDUSTRY BABY (feat. Jack Harlow)  33,948,454   

           Artist  Artist Followers                 Song ID  \
0        Måneskin         3377762.0  3Wrjm47oTz2sjIgck11l5e   
1   The Kid LAROI         2230022.0  5HCyWlXZPP0y6Gqq8TgA20   
2  Olivia Rodrigo         6266514.0  4ZtFanR9U6ndgddUvNcjcG   
3      Ed Sheeran        83293380.0  6PQ88X9TkUIAUIZJHW2upE   
4       Lil Nas X         5473565.0  27NovPIUIRrOZoCHxABJwK   

                                    Genre  ... Danceability Energy  Loudness  \
0  ['indie rock italiano', 'italian pop']  ...        0.714  0.800    -4.808   
1                  ['australian hip hop']  ...        0.591  0.764    -5.484   
2                                 ['pop']  ...        0.563  0.664    -5.044   
3                       ['pop', 'uk pop']  ...        0.808  0.897    -3.712   
4           ['lgbtq+ hip hop', 'pop rap']  ...        0.736  0.704    -7.409   

   Speechiness  Acousticness  Liveness    Tempo  Duration (ms)  Valence  Chord  
0       0.0504        0.1270    0.3590  134.002       211560.0    0.589      B  
1       0.0483        0.0383    0.1030  169.928       141806.0    0.478  C#/Db  
2       0.1540        0.3350    0.0849  166.928       178147.0    0.688      A  
3       0.0348        0.0469    0.3640  126.026       231041.0    0.591      B  
4       0.0615        0.0203    0.0501  149.995       212000.0    0.894  D#/Eb  

[5 rows x 23 columns]

In [161]:
top.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1556 entries, 0 to 1555
Data columns (total 23 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Index                      1556 non-null   int64  
 1   Highest Charting Position  1556 non-null   int64  
 2   Number of Times Charted    1556 non-null   int64  
 3   Week of Highest Charting   1556 non-null   object 
 4   Song Name                  1556 non-null   object 
 5   Streams                    1556 non-null   object 
 6   Artist                     1556 non-null   object 
 7   Artist Followers           1545 non-null   float64
 8   Song ID                    1545 non-null   object 
 9   Genre                      1545 non-null   object 
 10  Release Date               1545 non-null   object 
 11  Weeks Charted              1556 non-null   object 
 12  Popularity                 1545 non-null   float64
 13  Danceability               1545 non-null   float

In [162]:
top.columns = top.columns.str.lower().str.replace(' ', '_')
top.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1556 entries, 0 to 1555
Data columns (total 23 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   index                      1556 non-null   int64  
 1   highest_charting_position  1556 non-null   int64  
 2   number_of_times_charted    1556 non-null   int64  
 3   week_of_highest_charting   1556 non-null   object 
 4   song_name                  1556 non-null   object 
 5   streams                    1556 non-null   object 
 6   artist                     1556 non-null   object 
 7   artist_followers           1545 non-null   float64
 8   song_id                    1545 non-null   object 
 9   genre                      1545 non-null   object 
 10  release_date               1545 non-null   object 
 11  weeks_charted              1556 non-null   object 
 12  popularity                 1545 non-null   float64
 13  danceability               1545 non-null   float

In [163]:
top.rename(columns = {'artist': 'artist_name', "song_name": "track_name", "genre" : "genre_list", "chord": "key", "song_id": "track_id", "duration_(ms)": "duration_ms"}, inplace = True)
top.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1556 entries, 0 to 1555
Data columns (total 23 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   index                      1556 non-null   int64  
 1   highest_charting_position  1556 non-null   int64  
 2   number_of_times_charted    1556 non-null   int64  
 3   week_of_highest_charting   1556 non-null   object 
 4   track_name                 1556 non-null   object 
 5   streams                    1556 non-null   object 
 6   artist_name                1556 non-null   object 
 7   artist_followers           1545 non-null   float64
 8   track_id                   1545 non-null   object 
 9   genre_list                 1545 non-null   object 
 10  release_date               1545 non-null   object 
 11  weeks_charted              1556 non-null   object 
 12  popularity                 1545 non-null   float64
 13  danceability               1545 non-null   float

In [164]:
top.drop(columns = ['index'], inplace = True)
top["top_200"] = 1

In [165]:
## drop duplicates
top_dedup = top.drop_duplicates(subset = ['track_name', 'artist_name'], keep = 'first')
top_dedup.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1556 entries, 0 to 1555
Data columns (total 23 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   highest_charting_position  1556 non-null   int64  
 1   number_of_times_charted    1556 non-null   int64  
 2   week_of_highest_charting   1556 non-null   object 
 3   track_name                 1556 non-null   object 
 4   streams                    1556 non-null   object 
 5   artist_name                1556 non-null   object 
 6   artist_followers           1545 non-null   float64
 7   track_id                   1545 non-null   object 
 8   genre_list                 1545 non-null   object 
 9   release_date               1545 non-null   object 
 10  weeks_charted              1556 non-null   object 
 11  popularity                 1545 non-null   float64
 12  danceability               1545 non-null   float64
 13  energy                     1545 non-null   float

In [166]:
## Drop duplicates
db_dedup = db.drop_duplicates(subset = ['track_name', 'artist_name'], keep = 'first')
db_dedup.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 176514 entries, 0 to 232724
Data columns (total 18 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   genre             176514 non-null  object 
 1   artist_name       176514 non-null  object 
 2   track_name        176514 non-null  object 
 3   track_id          176514 non-null  object 
 4   popularity        176514 non-null  int64  
 5   acousticness      176514 non-null  float64
 6   danceability      176514 non-null  float64
 7   duration_ms       176514 non-null  int64  
 8   energy            176514 non-null  float64
 9   instrumentalness  176514 non-null  float64
 10  key               176514 non-null  object 
 11  liveness          176514 non-null  float64
 12  loudness          176514 non-null  float64
 13  mode              176514 non-null  object 
 14  speechiness       176514 non-null  float64
 15  tempo             176514 non-null  float64
 16  time_signature    17

In [167]:
db_dedup[db_dedup.duplicated(subset = ['track_name', 'artist_name'])].sort_values(by = ['track_name'])

Empty DataFrame
Columns: [genre, artist_name, track_name, track_id, popularity, acousticness, danceability, duration_ms, energy, instrumentalness, key, liveness, loudness, mode, speechiness, tempo, time_signature, valence]
Index: []

In [168]:
## Conservative dropping
db_semi_dedup = db.drop(columns= ['genre'])
db_semi_dedup.drop_duplicates(keep = 'first', inplace = True)
db_semi_dedup.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 191056 entries, 0 to 232724
Data columns (total 17 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   artist_name       191056 non-null  object 
 1   track_name        191056 non-null  object 
 2   track_id          191056 non-null  object 
 3   popularity        191056 non-null  int64  
 4   acousticness      191056 non-null  float64
 5   danceability      191056 non-null  float64
 6   duration_ms       191056 non-null  int64  
 7   energy            191056 non-null  float64
 8   instrumentalness  191056 non-null  float64
 9   key               191056 non-null  object 
 10  liveness          191056 non-null  float64
 11  loudness          191056 non-null  float64
 12  mode              191056 non-null  object 
 13  speechiness       191056 non-null  float64
 14  tempo             191056 non-null  float64
 15  time_signature    191056 non-null  object 
 16  valence           19

In [169]:
## View duplicates
db_semi_dedup[db_semi_dedup.duplicated(subset = ['track_name', 'artist_name'])].sort_values(by = ['track_name'])

artist_name                                         track_name  \
77209    Barns Courtney                                               "99"   
199606     Howard Shore                              "A Far Green Country"   
104937  Giacomo Puccini                                    "Nessun dorma!"   
100081       Can't Swim  "sometimes you meet the right people at the wr...   
204935     Yuki Hayashi                                                "上"   
...                 ...                                                ...   
205876     Yuki Hayashi                                                記録会   
206074     Yuki Hayashi                                            走るということ   
206385     Yuki Hayashi                                             踏み出す一歩   
130272     Joe Hisaishi                                             風のとおり道   
205697     Joe Hisaishi                                               風の伝説   

                      track_id  popularity  acousticness  danceability  \
77209   6YQUuoMnRIMaOmouYoMfQr          69      0.005980         0.552   
199606  1U5wqqEATmbCpSqADvSqy9          41      0.314000         0.218   
104937  74WjYdm3Lvbwnds4thYPUU          62      0.961000         0.171   
100081  1J5WWo4zSTWIKD7VFr6wn4          48      0.000109         0.382   
204935  5N5iImDfoQ4IF3YvXBpxPX          33      0.001230         0.255   
...                        ...         ...           ...           ...   
205876  2WGvTQyu6fy4b50rU5Ys8s          29      0.024500         0.394   
206074  6vzRpOC0JpkWIwd9M5N2a9          29      0.886000         0.414   
206385  7MnRIktherUlRJyn1iyZ4t          28      0.003760         0.458   
130272  0YQMQMXLWmbdXqKqkUjkWc          31      0.991000         0.436   
205697  6ObkZiJhsXuw4mbsXtETFo          30      0.987000         0.370   

        duration_ms  energy  instrumentalness key  liveness  loudness   mode  \
77209        199413  0.8040          0.000000  G#    0.1110    -4.299  Major   
199606        87920  0.0832          0.962000   C    0.2080   -23.025  Major   
104937       180933  0.3080          0.005460   D    0.3190   -12.383  Major   
100081       188994  0.8510          0.000004   E    0.3350    -6.023  Minor   
204935       241080  0.9540          0.920000   F    0.1110    -3.554  Major   
...             ...     ...               ...  ..       ...       ...    ...   
205876       185613  0.4380          0.777000   C    0.1100   -10.891  Minor   
206074       148333  0.1390          0.802000   E    0.1000   -16.356  Major   
206385       139667  0.5300          0.845000   A    0.1080    -9.171  Major   
130272       279232  0.1500          0.944000   G    0.0812   -17.835  Major   
205697       199804  0.1120          0.963000  G#    0.1120   -19.771  Major   

        speechiness    tempo time_signature  valence  
77209        0.0303   95.980            4/4   0.7140  
199606       0.0366   72.023            4/4   0.0795  
104937       0.0456  171.798            5/4   0.0889  
100081       0.0800  197.883            4/4   0.3310  
204935       0.2290  179.968            4/4   0.4060  
...             ...      ...            ...      ...  
205876       0.0401  100.045            4/4   0.2210  
206074       0.0296  139.888            4/4   0.0602  
206385       0.0274   89.960            4/4   0.5500  
130272       0.0494  149.995            4/4   0.2790  
205697       0.0479  152.932            4/4   0.1440  

[14542 rows x 17 columns]

In [170]:
db_semi_dedup[db_semi_dedup['artist_name'] == "Howard Shore"].sort_values(by = ['track_name'])

artist_name                                         track_name  \
199606  Howard Shore                              "A Far Green Country"   
125306  Howard Shore                              "A Far Green Country"   
125707  Howard Shore                "For Frodo" - feat. Ben del Maestro   
205252  Howard Shore  "Long Ways to Go Yet" / "Gollum's Song" (feat....   
205965  Howard Shore                "Where Is The Horse And The Rider?"   
...              ...                                                ...   
200964  Howard Shore                                      Wedding Plans   
128588  Howard Shore                                         Wilderland   
202718  Howard Shore                                         Wilderland   
124178  Howard Shore                                      Éowyn's Dream   
198629  Howard Shore                                      Éowyn's Dream   

                      track_id  popularity  acousticness  danceability  \
199606  1U5wqqEATmbCpSqADvSqy9          41         0.314        0.2180   
125306  1U5wqqEATmbCpSqADvSqy9          40         0.314        0.2180   
125707  4g7Ithd2Fpoupsjs5sJxeR          40         0.547        0.1640   
205252  7EUF4wSVqPUZfhpbUWWM65          30         0.934        0.0777   
205965  48b7rw6ibqzJ0Waw7yJ5Sr          28         0.897        0.0906   
...                        ...         ...           ...           ...   
200964  0M0QH0qRr27CeWiuBSP2YF          29         0.111        0.2240   
128588  4G3R4wcQnOYPqsI2Ujqp8J          26         0.346        0.0742   
202718  4G3R4wcQnOYPqsI2Ujqp8J          27         0.346        0.0742   
124178  6BEkaQDf8HOBD8woRzSmIH          44         0.972        0.2410   
198629  6BEkaQDf8HOBD8woRzSmIH          45         0.972        0.2410   

        duration_ms  energy  instrumentalness key  liveness  loudness   mode  \
199606        87920  0.0832           0.96200   C    0.2080   -23.025  Major   
125306        87920  0.0832           0.96200   C    0.2080   -23.025  Major   
125707       196867  0.3520           0.90900   E    0.0828   -14.595  Minor   
205252       485533  0.0708           0.12500   E    0.0908   -19.099  Minor   
205965       375680  0.0423           0.94400   G    0.1070   -23.446  Major   
...             ...     ...               ...  ..       ...       ...    ...   
200964       372107  0.5410           0.00478   C    0.1310    -8.633  Major   
128588       296413  0.2190           0.87400   A    0.1570   -17.794  Minor   
202718       296413  0.2190           0.87400   A    0.1570   -17.794  Minor   
124178        84307  0.0115           0.66200   A    0.0852   -33.576  Minor   
198629        84307  0.0115           0.66200   A    0.0852   -33.576  Minor   

        speechiness    tempo time_signature  valence  
199606       0.0366   72.023            4/4   0.0795  
125306       0.0366   72.023            4/4   0.0795  
125707       0.0369  130.973            4/4   0.1350  
205252       0.0438   60.650            4/4   0.0354  
205965       0.0430   87.089            4/4   0.0322  
...             ...      ...            ...      ...  
200964       0.0359  127.011            4/4   0.0934  
128588       0.0385   76.742            3/4   0.0359  
202718       0.0385   76.742            3/4   0.0359  
124178       0.0412   59.353            4/4   0.0374  
198629       0.0412   59.353            4/4   0.0374  

[330 rows x 17 columns]

Lots of songs in the 10000 per genre database were repeated in multiple genres. Additionally indexes (exa popularity) seemed to differ minorly between instances of songs. All duplicate artist + track name instance were dropped.

In [171]:
### Get columns to match exactly
print(top.columns)
db_dedup_match = db_dedup.drop(columns = ['genre', 'mode', 'instrumentalness', 'time_signature'])
db_dedup_match['top_200'] = 0
top_dedup_match = top_dedup.drop(columns = ['highest_charting_position', 'number_of_times_charted', 'week_of_highest_charting', 'streams', 'artist_followers', 'genre_list', 'release_date', 'weeks_charted'])


Index(['highest_charting_position', 'number_of_times_charted',
       'week_of_highest_charting', 'track_name', 'streams', 'artist_name',
       'artist_followers', 'track_id', 'genre_list', 'release_date',
       'weeks_charted', 'popularity', 'danceability', 'energy', 'loudness',
       'speechiness', 'acousticness', 'liveness', 'tempo', 'duration_ms',
       'valence', 'key', 'top_200'],
      dtype='object')


In [172]:
print(db_dedup_match.columns)
print(top_dedup_match.columns)
print(db_dedup_match.shape)
print(top_dedup_match.shape)

Index(['artist_name', 'track_name', 'track_id', 'popularity', 'acousticness',
       'danceability', 'duration_ms', 'energy', 'key', 'liveness', 'loudness',
       'speechiness', 'tempo', 'valence', 'top_200'],
      dtype='object')
Index(['track_name', 'artist_name', 'track_id', 'popularity', 'danceability',
       'energy', 'loudness', 'speechiness', 'acousticness', 'liveness',
       'tempo', 'duration_ms', 'valence', 'key', 'top_200'],
      dtype='object')
(176514, 15)
(1556, 15)


In [173]:
intersection = db_dedup.merge(top, on = ['artist_name', 'track_name'], how = 'inner')
intersection.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 151 entries, 0 to 150
Data columns (total 39 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   genre                      151 non-null    object 
 1   artist_name                151 non-null    object 
 2   track_name                 151 non-null    object 
 3   track_id_x                 151 non-null    object 
 4   popularity_x               151 non-null    int64  
 5   acousticness_x             151 non-null    float64
 6   danceability_x             151 non-null    float64
 7   duration_ms_x              151 non-null    int64  
 8   energy_x                   151 non-null    float64
 9   instrumentalness           151 non-null    float64
 10  key_x                      151 non-null    object 
 11  liveness_x                 151 non-null    float64
 12  loudness_x                 151 non-null    float64
 13  mode                       151 non-null    object 

In [174]:
merged_dups = pd.concat([db_dedup_match, top_dedup_match], axis = 0)
merged_dups.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 178070 entries, 0 to 1555
Data columns (total 15 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   artist_name   178070 non-null  object 
 1   track_name    178070 non-null  object 
 2   track_id      178059 non-null  object 
 3   popularity    178059 non-null  float64
 4   acousticness  178059 non-null  float64
 5   danceability  178059 non-null  float64
 6   duration_ms   178059 non-null  float64
 7   energy        178059 non-null  float64
 8   key           178059 non-null  object 
 9   liveness      178059 non-null  float64
 10  loudness      178059 non-null  float64
 11  speechiness   178059 non-null  float64
 12  tempo         178059 non-null  float64
 13  valence       178059 non-null  float64
 14  top_200       178070 non-null  int64  
dtypes: float64(10), int64(1), object(4)
memory usage: 21.7+ MB


In [175]:
merged_dedup = merged_dups.drop_duplicates(subset=['artist_name', 'track_name'], keep = "last")
merged_dedup.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 177919 entries, 0 to 1555
Data columns (total 15 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   artist_name   177919 non-null  object 
 1   track_name    177919 non-null  object 
 2   track_id      177908 non-null  object 
 3   popularity    177908 non-null  float64
 4   acousticness  177908 non-null  float64
 5   danceability  177908 non-null  float64
 6   duration_ms   177908 non-null  float64
 7   energy        177908 non-null  float64
 8   key           177908 non-null  object 
 9   liveness      177908 non-null  float64
 10  loudness      177908 non-null  float64
 11  speechiness   177908 non-null  float64
 12  tempo         177908 non-null  float64
 13  valence       177908 non-null  float64
 14  top_200       177919 non-null  int64  
dtypes: float64(10), int64(1), object(4)
memory usage: 21.7+ MB


In [176]:
merged_dedup.head()

artist_name                        track_name  \
0     Henri Salvador       C'est beau de faire un Show   
1  Martin & les fées  Perdu d'avance (par Gad Elmaleh)   
2    Joseph Williams    Don't Let Me Be Lonely Tonight   
3     Henri Salvador    Dis-moi Monsieur Gordon Cooper   
4       Fabien Nataf                         Ouverture   

                 track_id  popularity  acousticness  danceability  \
0  0BRjO6ga9RKCKjfDqeFgWV         0.0         0.611         0.389   
1  0BjC1NfoEOOusryehmNudP         1.0         0.246         0.590   
2  0CoSDzoNIKCRs124s9uTVy         3.0         0.952         0.663   
3  0Gc6TVm52BwZD07Ki6tIvf         0.0         0.703         0.240   
4  0IuslXpMROHdEPvSl1fTQK         4.0         0.950         0.331   

   duration_ms  energy key  liveness  loudness  speechiness    tempo  valence  \
0      99373.0   0.910  C#    0.3460    -1.828       0.0525  166.969    0.814   
1     137373.0   0.737  F#    0.1510    -5.559       0.0868  174.003    0.816   
2     170267.0   0.131   C    0.1030   -13.879       0.0362   99.488    0.368   
3     152427.0   0.326  C#    0.0985   -12.178       0.0395  171.758    0.227   
4      82625.0   0.225   F    0.2020   -21.150       0.0456  140.576    0.390   

   top_200  
0        0  
1        0  
2        0  
3        0  
4        0

In [186]:
merged_dedup.key.value_counts()

C        21071
G        20570
D        18732
A        17587
C#       16806
F        15705
E        13492
B        13004
A#       11896
F#       11336
G#       11142
D#        5940
C#/Db      214
G#/Ab      130
A#/Bb      122
F#/Gb      121
D#/Eb       40
Name: key, dtype: int64

In [187]:
# Bin similar keys together
merged_dedup['key'] = merged_dedup['key'].replace('C#/Db', 'C#')
merged_dedup['key'] = merged_dedup['key'].replace('G#/Ab', 'G#')
merged_dedup['key'] = merged_dedup['key'].replace('A#/Bb', 'A#')
merged_dedup['key'] = merged_dedup['key'].replace('F#/Gb', 'F#')
merged_dedup['key'] = merged_dedup['key'].replace('D#/Eb', 'D#')
merged_dedup.key.value_counts()

<ipython-input-187-a60a9aa68497>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  merged_dedup['key'] = merged_dedup['key'].replace('C#/Db', 'C#')
<ipython-input-187-a60a9aa68497>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  merged_dedup['key'] = merged_dedup['key'].replace('G#/Ab', 'G#')
<ipython-input-187-a60a9aa68497>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentat

C     21071
G     20570
D     18732
A     17587
C#    17020
F     15705
E     13492
B     13004
A#    12018
F#    11457
G#    11272
D#     5980
Name: key, dtype: int64

In [177]:
merged_dedup.to_csv("Resources/merged_dedup.csv", index = False)

In [180]:
# num_features = ['popularity', 'danceability', 'energy', 'loudness', 'speechiness', 'acousticness', 'liveness', 'valence', 'tempo', 'duration_ms']
# for col in num_features:
#     print(col)
#     merged_dedup[col] = merged_dedup[col].str.strip()

# merged_dedup.info()

Current dfs:
* db = raw data of 10000 songs per genre
* top = raw data of Top 200 charting songs for 2020-2021
    * cleaned top column names to match db
* top_dedup = removing all duplicate instances of the same track_name and artist_name (there are none)
* db_dedup = removing all duplicate instances of the same track_name and artist_name
* db_semi_dedup = ignoring genre, removing all duplicate rows with any different value (there are instances of the same song+artist combo with minorly different stats i.e. popularity 41 vs 42)
* intersection = 151 song+artist combos that appear in both dfs
* merged_dups = simple concat of two dataframes, the 151 in both appear twice
* merged_dedup = 151 duplicates removed, kept the second instance where they have top_200 == 1

## DATA PREPROCESSING

In [196]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [197]:
X = merged_dedup.drop(columns = ['artist_name', 'track_name', 'track_id', 'top_200'])
y = merged_dedup['top_200']

In [198]:
X_dummy = pd.get_dummies(X)
X_dummy.head()

popularity  acousticness  danceability  duration_ms  energy  liveness  \
0         0.0         0.611         0.389      99373.0   0.910    0.3460   
1         1.0         0.246         0.590     137373.0   0.737    0.1510   
2         3.0         0.952         0.663     170267.0   0.131    0.1030   
3         0.0         0.703         0.240     152427.0   0.326    0.0985   
4         4.0         0.950         0.331      82625.0   0.225    0.2020   

   loudness  speechiness    tempo  valence  ...  key_B  key_C  key_C#  key_D  \
0    -1.828       0.0525  166.969    0.814  ...      0      0       1      0   
1    -5.559       0.0868  174.003    0.816  ...      0      0       0      0   
2   -13.879       0.0362   99.488    0.368  ...      0      1       0      0   
3   -12.178       0.0395  171.758    0.227  ...      0      0       1      0   
4   -21.150       0.0456  140.576    0.390  ...      0      0       0      0   

   key_D#  key_E  key_F  key_F#  key_G  key_G#  
0       0      0      0       0      0       0  
1       0      0      0       1      0       0  
2       0      0      0       0      0       0  
3       0      0      0       0      0       0  
4       0      0      1       0      0       0  

[5 rows x 22 columns]

In [199]:
X_train, X_test, y_train, y_test = train_test_split(X_dummy, y, random_state=42)

In [200]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)